internal/controllers/machinedeployment/machinedeployment_inplace.go

```diff
@@ -3,42 +3,66 @@ package machinedeployment
 	"sigs.k8s.io/cluster-api/util/annotations"
 )
 
 func (r *Reconciler) rolloutInPlace(ctx context.Context, md *clusterv1.MachineDeployment, msList []*clusterv1.MachineSet) (reterr error) {
 	log := ctrl.LoggerFrom(ctx)
 
-	// For in-place upgrade, we shouldn't try to create a new MachineSet as that would trigger a rollout.
-	// Instead, we should try to get latest MachineSet that matches the MachineDeployment.Spec.Template/
+	// If there are no MachineSets for a MachineDeployment, either this is a create operation for a new
+	// MachineDeployment or the MachineSets were manually deleted. In either case, a new MachineSet should be created
+	// as there are no MachineSets that can be in-place upgraded.
+	// If there are already MachineSets present, we shouldn't try to create a new MachineSet as that would trigger a rollout.
+	// Instead, we should try to get latest MachineSet that matches the MachineDeployment.Spec.Template
 	// If no such MachineSet exists yet, this means the MachineSet hasn't been in-place upgraded yet.
 	// The external in-place upgrade implementer is responsible for updating the latest MachineSet's template
 	// after in-place upgrade of all worker nodes belonging to the MD is complete.
 	// Once the MachineSet is updated, this function will return the latest MachineSet that matches the
 	// MachineDeployment template and thus we can deduce that the in-place upgrade is complete.
-	newMachineSet, oldMachineSets, err := r.getAllMachineSetsAndSyncRevision(ctx, md, msList, false)
+	newMachineSetNeeded := len(msList) == 0
+	newMachineSet, oldMachineSets, err := r.getAllMachineSetsAndSyncRevision(ctx, md, msList, newMachineSetNeeded)
 	if err != nil {
 		return err
 	}
 
-	defer func() {
-		allMSs := append(oldMachineSets, newMachineSet)
-
-		// Always attempt to sync the status
-		err := r.syncDeploymentStatus(allMSs, newMachineSet, md)
-		reterr = kerrors.NewAggregate([]error{reterr, err})
-	}()
+	allMSs := append(oldMachineSets, newMachineSet)
 
 	if newMachineSet == nil {
 		log.Info("Changes detected, InPlace upgrade strategy detected, adding the annotation")
 		annotations.AddAnnotations(md, map[string]string{clusterv1.MachineDeploymentInPlaceUpgradeAnnotation: "true"})
-		return errors.New("new MachineSet not found. This most likely means that the in-place upgrade hasn't finished yet")
+	} else if !annotations.HasAnnotation(md, clusterv1.MachineDeploymentInPlaceUpgradeAnnotation) {
+		// If in-place upgrade annotation is no longer present, attempt to scale up the new MachineSet if necessary
+		// and scale down the old MachineSets if necessary.
+		// Note that if there are no scaling operations required, this else if block will be a no-op.
+
+		// Scale up, if we can.
+		if err := r.reconcileNewMachineSet(ctx, allMSs, newMachineSet, md); err != nil {
+			return err
+		}
+
+		if err := r.syncDeploymentStatus(allMSs, newMachineSet, md); err != nil {
+			return err
+		}
```

Comment:
We have this below at line 57, so do we need this here too? Or is it so that we can be more in sync? Before it was part of a defer, so even if another block returns an error, it would always sync, so wondering if we need it in a defer still

Reply:
I'm using similar logic as [here](https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/machinedeployment/machinedeployment_rolling.go#L48-L69) for create/scale case.